### PR TITLE
Add marlin int4 kernel

### PR DESCRIPTION
**Pull request by dacorvo**
# What does this PR do?

This adds a modified Marlin `fp16`/`int4` kernel to the library and creates two new `QTensor` subclasses to use it:

- `MarlinInt4PackedTensor`,
- `MarlinInt4WeightQBitsTensor`.

The AWQ kernel is still used by default because from the first tests it seems the modified Marlin kernel either has some accuracy issues or is not properly integrated (perplexity increases).

Note: during the integration, I tried to register the Marlin fp16int4 gemm as a `torch.tensor.library.custom_op`, but it added an extra latency (up to 50 %), so I used the same legacy declaration (with `define`/`impl`).

### CODE REVIEW

**Comment by dacorvo** [comment]
Closing this as the modified kernel is definitely flawed: as soon as it processes more than 32 inputs (i.e. two blocks of 16), there are errors in the outputs starting from the 128th output feature. There is likely some kind of flaw in the weight/scales/zero-point readback as soon as parallelization increases.
